README.md
task-logger.sh [![Build Status](https://travis-ci.org/posva/task-logger.sh.svg?branch=master)](https://travis-ci.org/posva/task-logger.sh)
===

![img](https://cloud.githubusercontent.com/assets/664177/7904349/fa16e226-07f7-11e5-91d5-7255b2c35930.gif)

Shell library to log tasks with nice output. Supports zsh and bash.

|      Property| Value                                                                       |
|-------------:|:----------------------------------------------------------------------------|
|      Version | 1.3.4                                                                       |
|       Author | Eduardo San Martin Morote                                                   |
|      License | The MIT License                                                             |
| Requirements | `perl` With Time::HiRes (comes by default in almost every Linux and in osx).|

# Example

```sh
#! /bin/bash

# Load the lib
source task-logger.sh

# Print some information message
info "Logs are available at $LOG_DIR"

# Print a message and do the corresponding message
# If the task fails, treat it as an error
working -n "Taking a nap"
log_cmd nap sleep 3 || ko

# Print a summary about what have been done
finish
```

Don't forget to call `finish` at the end (It is not mandatory, but summaries are
cool)!

#Motivation

When writing shell scripts you usually want to perform tasks with minimal output
while logging everything somewhere else in order to check it later.  Critical
tasks should stop the script when they fail. At the end you should have a
summary about errors, warnings and successes.

Everything with colors and unicode to make your shell look
:sparkles:fabulous:sparkles:.

#Features

* :lollipop:Colors
* Timers: every task displays the elapsed time
* Customizable:
  * Change any color
  * Create your own messages functions with 2 lines of code
  * Create your own progress loop (see [Working loop](#working-loop))
* Critical tasks
* Summary
* Can be killed
* Smart cleanup

#Usage

###Initialisation
As every shell lib you start by sourcing it: `source task-logger.sh`

###Messages
The lib has some printing functions defined. Every one of them follows this
structure:

```sh
warning() {
  parse_opt "$@"
  echo ${opts[@]} "${WARNING_COLOR}${args[@]}${RESET_COLOR}"
}
```

You can define your own functions adding more text and changing colors.
You can use `printf` or any other printing functions instead of `echo`.
See [Working loop](#working-loop) for an example.

The available functions are:

* `important`: prints a message with a blue background so it is quite visible
* `warning`: prints a message in yellow
* `good`: prints a message in green
* `bad`: prints a message in red
* `info`: prints a message in gray
* `error`: prints a message in white with red background. best fit for critical
    errors
* `ok`, `ko` and `warn` increments the numbers of successes, errors and
    warnings, respectively.  They also print fancy unicode characters
* `working`: prints the current time with format `[HH:MM:SS]` in `info` color
    and the prints a message in blue. This function is usually called with the
    `-n` option (no end line) and just before calling `log_cmd`.

###Logs
Launch tasks with `log_cmd`:

```sh
# log_cmd [-c] <task-name> <command> || (ko|warn)

log_cmd -c critical-task sleep 2 || ko
log_cmd error-task not-a-cmd || ko
log_cmd warn-task not-a-cmd || warn

# prints a summary
finish
```

###Cleanup
All of the logs are carefully saved in a temporary directory. This directory is
created when task-logger.sh is sourced and stored at `$LOG_DIR`.

When you call the `finish` method the temporary directory is cleaned up if there
were no errors. You can also pass the `--force-cleanup` option to clean the
`$LOG_DIR` anyways.  You can also manually cleanup by calling the `tmp_cleanup`.

If you need to create a new log dir use the `new_log_dir` method. This will
populate the `$LOG_DIR` variable but **will not** clean up the previous
directory. To do so you must either call the `finish` method (which will
eventually clean if there are no errors) or the `tmp_cleanup` method.

###Customization

A good example about how to customize the lib is in my
[blog](http://posva.net/shell/2015/02/03/using-task-loggersh/) where I use
task-logger.sh to log my dotfiles install script while keeping a nice output.

####Working loop

When I talk about working loop I refer to the loop called once a task is
launched. In the first version of `task-logger.sh` I was using `dot_working`,
which can still be used by setting the variable `WORKING` to `dot_working` and
`WORKING_END` to `true`. By default the lib use `turning_circle` and
`turning_circle_end`.  Creating your own functions is easy:

* One have an infinite loop and should call itself (`while true; do
...; done`)
* The other one is called when the loop is stopped. It can be used
to move the cursor backwards like I do in `turning_circle_end` using `printf
"\033[3D"`.

If you implement some cool working loop, please create a pull request so
everyone can enjoy it :smile:.

###Releases

You can find release history in CHANGELOG.md.

When updating the lib a new version must be released in a separated commit.
This commit must only contain the following additions:
* HISTORY.md: new release entry
* README.md: update version
* task-logger.sh: update version and date

##TODO

* improve `cygwin` compatibility. It does work at the moment except for the
    missing unicode characters. Ctrl+C is printing as it should
* Use some kind of template for working loops and share some variables with it
    from the `log_cmd` method



HISTORY.md
# 1.3.3 - 05/08/2015
* Fixed bug with get_timer

# 1.3.2 - 03/08/2015
* Fixed bug with log_cmd that prevented command with options to work as
    expexted

# 1.3.1 - 29/07/2015
* Added better elapsed time display with [pretty-hrtime.sh]
    (https://github.com/posva/pretty-hrtime.sh)
* Removed unused local variables
* Removed bc dependency :relieved:

# 1.3.0 - 27/07/2015
* Added smart cleanup
    * Added options to the finish method
* `tmp_cleanup` method
* Added history.md

# 1.2.0 - 01/06/2015
* Added overwrite option in `log_task`
* Added last modification date in script

# 1.1.0 - 06/02/2015
* Added a better in progress indicator
* Ability to customize the progress indicator
* Actually fixed bug with negative elapsed time

# 1.0.1 - 02/02/2015
* Fixed bug with negative elapsed time
* Don't overwrite tasks with the same name. Instead add an increasing number at
    the end

# 1.0.0 - 30/01/2015
* Initial release


test.sh
#! /bin/bash

error=
for f in tests/*; do
  if ! ./"$f"; then
    error=YES
  fi
done

if [[ "$error" == YES ]]; then
  exit 1
fi


all-tests.sh
#! /bin/bash

if [[ "$1" == "" ]]; then
  shell=bash
elif grep "$1" /etc/shells 2>/dev/null 1>/dev/null; then
  shell="$1"
  shift
else
  shell=bash
fi
if [[ "$1" == "" ]]; then
  files="tests/*.sh"
else
  files="$@"
fi

if [[ "$shell" == "zsh" ]];then
  shell="zsh -y"
fi

error=
mkdir -p work
for f in $files; do
  n=work/$(basename "$f")
  sed "1s/.*/#!\/bin\/${shell}/" "$f" > "$n"
  chmod +x "$n"
  if ! ./"$n"; then
    error=YES
  fi
done

if [[ "$error" == YES ]]; then
  exit 1
fi


tests/all.sh
#!/bin/bash
#file: tests.sh

source task-logger.sh

testCleanIfNoErrors() {
  no_colors
  log_cmd test true > /dev/null || ko > /dev/null
  assertTrue "[[ -d "$LOG_DIR" && -f "$LOG_DIR/test.out" ]]"
  finish > /dev/null
  assertTrue "[[ ! -d "$LOG_DIR" && ! -f "$LOG_DIR/test.out" ]]"
  new_log_dir
  reset_counters
}

testNoCleanIfErrors() {
  no_colors
  log_cmd test false > /dev/null || ko > /dev/null
  assertTrue "[[ -d "$LOG_DIR" && -f "$LOG_DIR/test.out" ]]"
  finish > /dev/null
  assertTrue "[[ -d "$LOG_DIR" && -f "$LOG_DIR/test.out" ]]"
  tmp_cleanup
  new_log_dir
  reset_counters
}

testNoCleanIfOption() {
  no_colors
  log_cmd test true > /dev/null || ko > /dev/null
  assertTrue "[[ -d "$LOG_DIR" && -f "$LOG_DIR/test.out" ]]"
  finish --no-cleanup > /dev/null
  assertTrue "[[ -d "$LOG_DIR" && -f "$LOG_DIR/test.out" ]]"
  tmp_cleanup
  new_log_dir
  reset_counters
}

testCleanIfForce() {
  no_colors
  log_cmd test false > /dev/null || ko > /dev/null
  assertTrue "[[ -d "$LOG_DIR" && -f "$LOG_DIR/test.out" ]]"
  finish --force-cleanup > /dev/null
  assertTrue "[[ ! -d "$LOG_DIR" && ! -f "$LOG_DIR/test.out" ]]"
  new_log_dir
  reset_counters
}

# first with colors
testColors() {
  set_colors
  assertEquals "${INFO_COLOR}sample text${RESET_COLOR}" "$(info "sample text")"
  assertEquals "${GOOD_COLOR}sample text${RESET_COLOR}" "$(good "sample text")"
  assertEquals "${BAD_COLOR}sample text${RESET_COLOR}" "$(bad "sample text")"
  assertEquals "${WARNING_COLOR}sample text${RESET_COLOR}" "$(warning "sample text")"
  assertEquals "${ERROR_COLOR}sample text${RESET_COLOR}" "$(error "sample text")"
  assertEquals "${IMPORTANT_COLOR}sample text${RESET_COLOR}" "$(important "sample text")"
  assertEquals "${INFO_COLOR}[00:00:00] ${RESET_COLOR}${WORKING_COLOR}sample text${RESET_COLOR}" "$(working "sample text" | sed "s/\[[0-9]*:[0-9]*:[0-9]*\]/[00:00:00]/g")"
}

#then without
testNoColors() {
  no_colors
  assertEquals "sample text" "$(info "sample text")"
  assertEquals "sample text" "$(good "sample text")"
  assertEquals "sample text" "$(bad "sample text")"
  assertEquals "sample text" "$(warning "sample text")"
  assertEquals "sample text" "$(error "sample text")"
  assertEquals "sample text" "$(important "sample text")"
  assertEquals "[00:00:00] sample text" "$(working "sample text" | sed "s/\[[0-9]*:[0-9]*:[0-9]*\]/[00:00:00]/g")"
}

testCounters() {
  # easier testing
  no_colors

  assertEquals "0" "$(echo ${SUCCESS})"
  assertEquals "0" "$(echo ${WARNINGS})"
  assertEquals "0" "$(echo ${ERRORS})"
  assertEquals " ${SUCCESS_SYMBOL} " "$(ok)"
  ok > /dev/null
  assertEquals "1" "$(echo ${SUCCESS})"
  assertEquals "0" "$(echo ${WARNINGS})"
  assertEquals "0" "$(echo ${ERRORS})"
  assertEquals " ${ERROR_SYMBOL} " "$(ko)"
  ko > /dev/null
  assertEquals "1" "$(echo ${SUCCESS})"
  assertEquals "0" "$(echo ${WARNINGS})"
  assertEquals "1" "$(echo ${ERRORS})"
  assertEquals " ${WARNING_SYMBOL} " "$(warn)"
  warn > /dev/null
  assertEquals "1" "$(echo ${SUCCESS})"
  assertEquals "1" "$(echo ${WARNINGS})"
  assertEquals "1" "$(echo ${ERRORS})"


  ok > /dev/null
  ok > /dev/null
  assertEquals "3" "$(echo ${SUCCESS})"
  assertEquals "1" "$(echo ${WARNINGS})"
  assertEquals "1" "$(echo ${ERRORS})"

  warn > /dev/null
  assertEquals "3" "$(echo ${SUCCESS})"
  assertEquals "2" "$(echo ${WARNINGS})"
  assertEquals "1" "$(echo ${ERRORS})"

  ko > /dev/null
  ko > /dev/null
  ko > /dev/null
  ko > /dev/null
  assertEquals "3" "$(echo ${SUCCESS})"
  assertEquals "2" "$(echo ${WARNINGS})"
  assertEquals "5" "$(echo ${ERRORS})"
}

custom_fun() {
  echo "normal output"
  echo "error output" >&2
}

testWorking() {
  # easier testing
  no_colors

  # test with another WORKING
  WORKING=turning_circle
  WORKING_END=turning_circle_end

  OUT=$(log_cmd custom custom_fun | sed 's/.*\[3D.*\[3D *\[[0-9.][0-9.]* [mns]*\] '${SUCCESS_SYMBOL}' /YES/')

  assertEquals "YES" "$OUT"
  assertEquals "normal output" "$(cat ${LOG_DIR}/custom.out)"
  assertEquals "error output" "$(cat ${LOG_DIR}/custom.err)"

  OUT=$( (log_cmd fail idontexists || ko) | sed 's/.*\[3D.*\[3D *\[[0-9.][0-9.]* [mns]*\] '${ERROR_SYMBOL}' /YES/')

  assertEquals "YES" "$OUT"
}

error_test() {
  echo lol >&2
  log_cmd error bad-cmd || ko
  echo end >&2
}
warn_test() {
  log_cmd warn bad-cmd || warn
}

testLog() {
  # easier testing
  no_colors
  WORKING=dot_working
  WORKING_END=true

  assertEquals "YES" "$(echo $LOG_DIR | sed 's/.*task-logger.*/YES/')"

  assertEquals "YES" "$(log_cmd task echo Hello | sed 's/\.\.*\[[0-9.][0-9.]* [mns]*\] '${SUCCESS_SYMBOL}' /YES/')"
  assertEquals "Hello" "$(cat ${LOG_DIR}/task.out)"

  assertEquals "YES" "$(log_cmd custom custom_fun | sed 's/\.\.*\[[0-9.][0-9.]* [mns]*\] '${SUCCESS_SYMBOL}' /YES/')"
  assertEquals "normal output" "$(cat ${LOG_DIR}/custom.out)"
  assertEquals "error output" "$(cat ${LOG_DIR}/custom.err)"

  # two tasks with the same name
  assertEquals "YES" "$(log_cmd task echo Hello | sed 's/\.\.*\[[0-9.][0-9.]* [mns]*\] '${SUCCESS_SYMBOL}' /YES/')"
  assertEquals "Hello" "$(cat ${LOG_DIR}/task-1.out)"

  assertEquals "YES" "$(log_cmd task echo Hello | sed 's/\.\.*\[[0-9.][0-9.]* [mns]*\] '${SUCCESS_SYMBOL}' /YES/')"
  assertEquals "Hello" "$(cat ${LOG_DIR}/task-2.out)"
}

testMessages() {
  # no colors for easier testing
  no_colors

  #Works without quotes
  assertEquals "message without quotes" "$(info message without quotes)"
  assertEquals "message without quotes" "$(good message without quotes)"
  assertEquals "message without quotes" "$(bad message without quotes)"
  assertEquals "message without quotes" "$(error message without quotes)"
  assertEquals "message without quotes" "$(warning message without quotes)"
  assertEquals "message without quotes" "$(important message without quotes)"
  assertEquals "[00:00:00] message without quotes" "$(working message without quotes | sed "s/\[[0-9]*:[0-9]*:[0-9]*\]/[00:00:00]/g")"

  # test -n for echo
  assertEquals "message without quotes OK" "$(info -n message without quotes; echo " OK")"
  assertEquals "message without quotes OK" "$(good -n message without quotes; echo " OK")"
  assertEquals "message without quotes OK" "$(bad -n message without quotes; echo " OK")"
  assertEquals "message without quotes OK" "$(error -n message without quotes; echo " OK")"
  assertEquals "message without quotes OK" "$(warning -n message without quotes; echo " OK")"
  assertEquals "message without quotes OK" "$(important -n message without quotes; echo " OK")"
  assertEquals "[00:00:00] message without quotes OK" "$((working -n message without quotes; echo " OK") | sed "s/\[[0-9]*:[0-9]*:[0-9]*\]/[00:00:00]/g")"
}

testReturnCodes() {
  no_colors

  # symbols
  assertFalse 'ko 1 >/dev/null 2>/dev/null'
  assertTrue 'ok >/dev/null 2>/dev/null'
  assertTrue 'warn >/dev/null 2>/dev/null'

  # message helpers
  assertFalse 'error critical error >/dev/null 2>/dev/null'
  assertTrue 'info >/dev/null 2>/dev/null'
  assertTrue 'good >/dev/null 2>/dev/null'
  assertTrue 'bad >/dev/null 2>/dev/null'
  assertTrue 'warning >/dev/null 2>/dev/null'
  assertTrue 'working >/dev/null 2>/dev/null'
  assertTrue 'important >/dev/null 2>/dev/null'

  # log_cmd
  log_cmd task-name non-existant-command >/dev/null 2>/dev/null
  assertEquals "$?" 127
  assertFalse 'log_cmd grep grep unexistant-word-in-lib task-logger.sh >/dev/null 2>/dev/null'
  assertTrue 'log_cmd echo echo >/dev/null 2>/dev/null'
}

testFinish() {
  no_colors

  ERRORS=3
  SUCCESS=2
  WARNINGS=34
  assertEquals "[00:00:00] Finished: $SUCCESS ✓ $WARNINGS ⚠ $ERRORS ✗" "$(finish | sed -e "s/\[[0-9]*:[0-9]*:[0-9]*\]/[00:00:00]/g" -e 's/[^✗]*$//')"
}

testCritical() {
  no_colors

  # non failing
  assertTrue 'log_cmd -c task-name echo >/dev/null 2>/dev/null'

  # failing
  # I must quit the less loop somehow
  #assertFalse 'log_cmd -c task-name nope'
}

testOverwrite() {
  no_colors

  log_cmd -o overwrite 'echo overwritetest' >/dev/null 2>/dev/null
  assertTrue "[[ -f '$LOG_DIR/overwrite.out' && -f '$LOG_DIR/overwrite.out' ]]"
  assertEquals "overwritetest" "$(cat $LOG_DIR/overwrite.out)"

  log_cmd -o overwrite 'echo overwritetest2' >/dev/null 2>/dev/null
  assertTrue "[[ -f '$LOG_DIR/overwrite.out' && -f '$LOG_DIR/overwrite.err' ]]"
  assertTrue "[[ ! -f '$LOG_DIR/overwrite-1.out' && ! -f '$LOG_DIR/overwrite-1.err' ]]"
  assertEquals "overwritetest2" "$(cat $LOG_DIR/overwrite.out)"
}

testLogWithOptions() {
  no_colors

  assertEquals "$(log_cmd task echo -n Hello | sed 's/\.\.*\[[0-9.][0-9.]* [mns]*\] '${SUCCESS_SYMBOL}' /YES/')" "YES"
  assertEquals "$(cat $LOG_DIR/task.out)" "Hello"
}

# Must be called at the end
# It cleans up the temporary dirs used
testTeardown() {
  tmp_cleanup
  show_cursor
}

SHUNIT_PARENT="$0"
source lib/shunit2


task-logger.sh
#! /bin/bash

# task-logger.sh 1.3.4 - zsh and bash fancy task logging
# Copyright (C) 2015 Eduardo San Martin Morote
# Last modification 2015-08-11
#
# http://github.com/posva/task-logger.sh
#
# Permission is hereby granted, free of charge, to any person obtaining a copy of
# this software and associated documentation files (the "Software"), to deal in
# the Software without restriction, including without limitation the rights to
# use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
# of the Software, and to permit persons to whom the Software is furnished to do
# so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

# Counting for errors
ERRORS=0
# Counting for warnings
WARNINGS=0
# Counting for successes
SUCCESS=0

SUCCESS_SYMBOL="✓"
ERROR_SYMBOL="✗"
WARNING_SYMBOL="⚠"
KILLED_SYMBOL="☠"

# Set color variables. They can be overrided.
set_colors() {
  RESET_COLOR="[0m"
  IMPORTANT_COLOR="[104;30m"
  WARNING_COLOR="[33m"
  GOOD_COLOR="[32m"
  BAD_COLOR="[31m"
  INFO_COLOR="[90m"
  ERROR_COLOR="[101m"
  WORKING_COLOR="[94m"
  END_GOOD_COLOR="[92m"
  END_BAD_COLOR="[91m"
  END_WARNING_COLOR="[93m"
}

# Unset colors
no_colors() {
  RESET_COLOR=
  IMPORTANT_COLOR=
  WARNING_COLOR=
  GOOD_COLOR=
  BAD_COLOR=
  INFO_COLOR=
  ERROR_COLOR=
  WORKING_COLOR=
  END_GOOD_COLOR=
  END_BAD_COLOR=
  END_WARNING_COLOR=
}

show_cursor() {
  echo -n "[?25h"
}

hide_cursor() {
  echo -n "[?25l"
}

# https://github.com/posva/pretty-hrtime.sh
_desc=(d:86400:day h:3600:hour m:60:minute s:1:second ms:1000000:millisecond μs:1000:microsecond ns:1:nanosecond)
ptime() {
  local result s ns amount unit convert lconvert verbose tmp_str long
  for i in "$@"; do
    case $i in
      -v|--verbose)
        verbose=YES
        shift
        ;;
      -l|--long)
        long=YES
        shift
        ;;
      -[a-zA-Z]|--*)
        echo "[ptime] Option $i doesn't exists" >&2
        shift
        ;;
    esac
  done
  s="$1"
  ns="$2"
  for desc in ${_desc[@]}; do
    convert=$(echo "$desc" | cut -d: -f2)
    unit=$(echo "$desc" | cut -d: -f1)
    # Use the seconds or the nanoseconds amount
    if echo "$unit" | grep '.s$' > /dev/null; then
      amount="$ns"
    else
      amount="$s"
    fi
    # Remove any extra time already counted
    if [[ "$unit" != d && "$unit" != ms ]]; then
      (( amount %= lconvert ))
    fi
    (( val = (100 * amount) /  convert ))
    if [[ "$unit" = s ]]; then
      (( val += ns / 10000000 ))
    fi
    if [[ "$val" -ge 100 ]]; then
      # Don't print to many decimals
      if [[ "$long" = YES || "$val" -ge 1000 ]]; then
        (( val /= 100 ))
        tmp_str="$val"
      else
        # We need a tmp var because val is treated as a number
        tmp_str=$(echo "$val" | sed -e 's/..$/.&/' -e 's/^.$/.0&/' -e 's/\.0*$//' -e 's/00*$//')
        if [[ -z "$tmp_str" ]]; then
          tmp_str=0
        fi
      fi
      if [[ -n "$result" ]]; then
        result="$result "
      fi
      if [[ "$verbose" = YES ]]; then
        unit=$(echo "$desc" | cut -d: -f3)
      fi
      result="${result}${tmp_str} $unit"
      if [[ "$verbose" = YES && "$tmp_str" != 1 ]]; then
        result="${result}s"
      fi
      if [[ -z "$long" ]]; then
        break
      fi
    fi
    # Save value for next iteration
    lconvert="$convert"
  done

  if [[ -z "$result" ]]; then
    if [[ -z "$verbose" ]]; then
      echo "0 s"
    else
      echo "0 seconds"
    fi
  else
    echo "$result"
  fi
}

# Parse options and args and place them in two global variables
# args and opts
# INTERNAL FUNCTION
parse_opt() {
  local io ia
  args=() # array
  opts=()
  io=1
  ia=1
  while [[ "$#" > 0 ]]; do #[[ "$1" =~ "$re" ]]; do
    if [[ "$1" =~ $re ]]; then
      opts[$io]=$1
      ((io++))
    else
      args[$ia]="$1"
      ((ia++))
    fi
    shift
  done
}

# WORKING should be some kind of loop. It is executed whenever a command
# is launched and can be customized. By default there are two options:
# dot_working and turning_circle. The last one being the default. To use your
# own loop just set WORKING to the function or program. Additionally you must
# also set WORKING_END to anything you want to call after the main loop is
# killed. For example I use to replace the cursor. You can set it to true or
# any other command that do not print anything
WORKING=turning_circle
WORKING_END=turning_circle_end

# How many seconds does a dot represent
DOT_SECONDS=1
# Print dots in order to show progress.
# This functions should be called with & and killed when job is done
# If you want to use this function, set WORKING to dot_working and
# WORKING_END to true
dot_working() {
  while true; do
    echo -n '.'
    sleep $DOT_SECONDS
  done
}

# Prints a turning circle with unicode to show work is in progress
turning_circle() {
  local p n symbols
  p=1
  n=4
  symbols=()
  symbols[1]=" ◐ "
  symbols[2]=" ◓ "
  symbols[3]=" ◑ "
  symbols[4]=" ◒ "

  #trap 'printf "\033[5D "; return' SIGINT
  #trap 'printf "\033[3D "; return' SIGHUP SIGTERM

  printf "   "
  while true; do
    printf "\033[3D${symbols[$p]}"
    ((p++))
    if [[ "$p" > "$n" ]]; then
      # :nocov:
      p=1
      # :nocov:
    fi
    sleep 0.2
  done
}

turning_circle_end() {
  printf "\033[3D "
}

# Check if perl is available
if perl -e 'use Time::HiRes qw( gettimeofday );' 2>&1 >/dev/null; then
  PERL=YES
fi

TIMER_INIT=()
if [[ -n "$PERL" ]]; then
  # Little timer helper using perl to have microseconds precision even in OS X
  # ex reset_timer 1 # set timer with id 1 at 0s
  reset_timer() {
    TIMER_INIT[$1]=$(perl -e 'use Time::HiRes qw( gettimeofday ); my ($a, $b) = gettimeofday; $ts = $a; $tn = $b * 1000; print "$ts $tn";')
  }

  # get the current value of a timer without resetting it
  # ex get_timer 1 get elapsed time since last reset_timer 1 call
  get_timer() {
    local elapsed seconds nanoseconds
    seconds=$(echo "${TIMER_INIT[$1]}" | cut -d ' ' -f 1)
    nanoseconds=$(echo "${TIMER_INIT[$1]}" | cut -d ' ' -f 2)
    elapsed=$(perl -e 'use Time::HiRes qw( gettimeofday ); my ($a, $b) = gettimeofday; $ts = $a - '"$seconds"'; $tn = $b * 1000 - '"$nanoseconds"'; if($ts == 1 && $tn < 0) {$tn = $ts * 100000000 - $tn; $ts = 0;} print "$ts $tn";')
    echo "$elapsed"
  }
elif date '+%s %N' 2>&1 >/dev/null; then
  reset_timer() {
    TIMER_INIT[$1]=$(date '+%s %N' | sed 's/ 00*/ /')
  }
  get_timer() {
    local elapsed seconds nanoseconds
    seconds=$(echo "${TIMER_INIT[$1]}" | cut -d ' ' -f 1)
    nanoseconds=$(echo "${TIMER_INIT[$1]}" | cut -d ' ' -f 2)
    ((
      seconds = $(date '+%s') - seconds,
      nanoseconds = $(date '+%N' | sed 's/^00*//') - nanoseconds
    ))
    if [[ "$seconds" = 1 && "$nanoseconds" < 0 ]]; then
      ((
        nanoseconds = seconds * 100000000 - nanoseconds,
        seconds = 0
      ))
    fi
    echo "$seconds $nanoseconds"
  }
else
  reset_timer() {
    TIMER_INIT[$1]="$(date '+%s')"
  }
  get_timer() {
    local elapsed seconds nanoseconds
    seconds="${TIMER_INIT[$1]}"
    nanoseconds=0
    (( seconds = $(date '+%s') - seconds ))
    if [[ "$seconds" < 0 ]]; then
      seconds=0
    fi
    echo "$seconds $nanoseconds"
  }
fi

# simple regex to parse options
re='^--?[a-zA-Z0-9]+'

# Important message that must be read and contains useful information
# Accepts same options as echo
important() {
  parse_opt "$@"
  echo ${opts[@]} "${IMPORTANT_COLOR}${args[@]}${RESET_COLOR}"
}

# Warning message in yellow
# Accepts same options as echo
warning() {
  parse_opt "$@"
  echo ${opts[@]} "${WARNING_COLOR}${args[@]}${RESET_COLOR}"
}

# Print a message in green
# Accepts same options as echo
good() {
  parse_opt "$@"
  echo ${opts[@]} "${GOOD_COLOR}${args[@]}${RESET_COLOR}"
}

# Print a message in red
# Accepts same options as echo
bad() {
  parse_opt "$@"
  echo ${opts[@]} "${BAD_COLOR}${args[@]}${RESET_COLOR}"
}

# Print a message in gray
# Accepts same options as echo
info() {
  parse_opt "$@"
  echo ${opts[@]} "${INFO_COLOR}${args[@]}${RESET_COLOR}"
}

# Error message with Red background
# Fits critical errors
# Accepts same options as echo
# Always return 1 (error)
error() {
  parse_opt "$@"
  echo ${opts[@]} "${ERROR_COLOR}${args[@]}${RESET_COLOR}"
  return 1
}

# Simple check mark. Increment the number of successes
ok() {
  ((SUCCESS++))
  echo "${GOOD_COLOR} ${SUCCESS_SYMBOL} ${RESET_COLOR}"
}

# Simple cross mark. Increment the number of errors
# Always return 1 (error)
ko() {
  echo "${BAD_COLOR} ${ERROR_SYMBOL} ${RESET_COLOR}"
  ((ERRORS++))
  return 1
}

# Simple warning mark. Increment the number of warnings
warn() {
  ((WARNINGS++))
  echo "${WARNING_COLOR} ${WARNING_SYMBOL} ${RESET_COLOR}"
}

# Helper message that prints time in gray and a message in blue
# This should be used before log_cmd
working() {
  info -n "[$(date +%H:%M:%S)] "
  parse_opt "$@"
  echo ${opts[@]} "${WORKING_COLOR}${args[@]}${RESET_COLOR}"
}

# Stops the dot_working function and print elapsed time with a mark depending
# whether the previously ran command ended correctly.
# INTERNAL FUNCTION
cleanup() {
  local elapsed
  while [[ "$DOT" == "" ]]; do
    # :nocov:
    sleep 1
    # :nocov:
  done
  kill $DOT 2>/dev/null
  wait $DOT 2>/dev/null
  DOT=
  if [[ "$1" != -99 ]]; then
    $WORKING_END
  fi
  elapsed="$(get_timer 1)"
  echo -n "[$(ptime $(echo "$elapsed"))]"
  if [[ "$1" == 0 ]]; then
    ok
  fi
}

# Function called when the user kills the script
killed() {
  # clean ^C
  printf "\033[2D"
  #kill 0
  if [[ "$DOT" != "" ]]; then
    cleanup -99
  fi
  bad " ${KILLED_SYMBOL} "
  finish
  exit 1
}

# Main command to run a task and log the output
# log_cmd [OPTIONS] NAME COMMAND || (ko|warn)
# Enter the command normally. You must escape any special shell characters (&, |, &&, ||, ;)
# Follow this command by || ko or || warn depending of how bad it is
# for the command to fail
# OPTIONS:
# -c, --critical: If this task is marked as critical, the scripto will stop
# right away and launch less on the last error output of the file
# -o, --overwrite: Don't use a new name for error and standard output files if
# the file already exists. You should always use this option if you don't need
# to save the output of a command ran inside a loop.
log_cmd() {
  local cmd critical p name p_cmd i overwrite tmp
  reset_timer 1
  $WORKING &
  DOT=$!
  critical=
  overwrite=
  for i in "$@"; do
    case $i in
      -c|--critical)
        critical=YES
        shift
        ;;
      -o|--overwrite)
        overwrite=YES
        shift
        ;;
      -[a-zA-Z0-9]|--*)
        error "[task-logger] Option $i doesn't exists for log_cmd"
        shift
        ;;
      *)
        # We got to the command
        break
      ;;
    esac
  done
  name="$1"
  # check if name can be be used as a file
  i=0
  if [[ -z "$overwrite" ]]; then
    tmp="$name"
    while [[ -f "${LOG_DIR}/${name}.out" || -f "${LOG_DIR}/${name}.err" ]]; do
      ((i++))
      name="$tmp-$i"
    done
  fi
  cmd="$2"
  shift
  shift
  if [[ "$critical" == YES ]]; then
    $cmd "$@" > ${LOG_DIR}/${name}.out 2> ${LOG_DIR}/${name}.err &
    p_cmd=$!
    wait $p_cmd 2>/dev/null
    p=$?
    cleanup $p
    if [[ "$p" != 0 ]]; then
      error ' CRITICAL '
      warning "Log is at ${LOG_DIR}/${name}.err"
      less ${LOG_DIR}/${name}.err
      exit 1
    fi
  else
    $cmd "$@" > ${LOG_DIR}/${name}.out 2> ${LOG_DIR}/${name}.err &
    p_cmd=$!
    wait $p_cmd 2>/dev/null
    p=$?
    cleanup $p
    return $p
  fi
}

# Helper function that should be called at the end.
# Prints a summary with the number of errors, warnings and successes
# If there are no errors, cleanup the whole log directory
finish() {
  local no_cleanup force_cleanup
  for i in "$@"; do
    case $i in
      -c|--no-cleanup)
        no_cleanup=YES
        shift
        ;;
      -f|--force-cleanup)
        force_cleanup=YES
        shift
        ;;
      *)
        ;;
    esac
  done
  info -n "[$(date +%H:%M:%S)] "
  echo "Finished: ${END_GOOD_COLOR}$SUCCESS ✓ ${END_WARNING_COLOR}$WARNINGS ⚠ ${END_BAD_COLOR}$ERRORS ✗${RESET_COLOR}"
  if test "$force_cleanup" -o \( "$ERRORS" -le 0 -a ! "$no_cleanup" \); then
    tmp_cleanup
  fi
  show_cursor
}

# Clean the tmp data. If your script run for a very long time you migth want
# to clean up the mess in tmp.
# This function may be called at any time and it is called by the finish
# method if there were no errors
tmp_cleanup() {
  local dir
  if [[ -n "$1" ]]; then
    dir="$1"
  else
    dir="$LOG_DIR"
  fi
  if ! rm -rf "$dir"; then
    bad "Error cleaning up the logs"
  fi
}

# Exit correctly with <C-C>
trap 'killed' SIGINT SIGTERM
set_colors

# Create a folder to redirect standard an error output
new_log_dir() {
  LOG_DIR=$(mktemp -d /tmp/task-logger-XXXXXXXX)
}
new_log_dir
hide_cursor

# Reset global variables used for counting errors, warnings and successes
# If you're calling finish multiple times you may need this
reset_counters() {
  ERRORS=0
  WARNINGS=0
  SUCCESS=0
}

